corehq/apps/sms/tasks.py

```diff
@@ -154,12 +181,21 @@ def handle_outgoing(msg):
             # Requeue the message and try it again shortly
             return True
 
+    if max_simultaneous_connections:
+        connection_slot = reserve_connection_slot(backend, max_simultaneous_connections)
+        if not connection_slot:
+            # Requeue the message and try it again shortly
+            return True
+
     result = send_message_via_backend(
         msg,
         backend=backend,
         orig_phone_number=orig_phone_number
     )
 
+    if max_simultaneous_connections:
+        free_connection_slot(backend, connection_slot)
```

Comment:
Flagging that if there's an exception in sending the message, the slot won't be freed up until it hits the timeout.  You could try/catch the in between calls or use a construct like `with reserve_connection_slot(...)` to guarantee that this gets called.

Reply:
Actually I guess `send_message_via_backend` is itself wrapped in a big try/except, huh?  Given that, this is probably safe enough to to do as-is.